xline/src/storage/compact/periodic_compactor.rs

```diff
@@ -0,0 +1,305 @@
+    ) -> Option<i64> {
+        if !self.is_leader.load(Relaxed) {
+            return None;
+        }
+        let target_revision = revision_window.expired_revision();
+        if target_revision == last_revision {
+            return None;
+        }
+        let revision =
+            target_revision.unwrap_or_else(|| unreachable!("target revision shouldn't be None"));
+        let now = Instant::now();
+        info!(
+            "starting auto periodic compaction, revision = {}, period = {:?}",
+            revision, self.period
+        );
+        // TODO: add more error processing logic
+        if let Err(e) = self.client.compact(revision).await {
+            warn!(
+                "failed auto revision compaction, revision = {}, period = {:?}, error: {:?}",
+                revision, self.period, e
+            );
+            None
+        } else {
+            info!(
+                "completed auto revision compaction, revision = {}, period = {:?}, took {:?}",
+                revision,
+                self.period,
+                now.elapsed().as_secs()
+            );
+            target_revision
+        }
+    }
+}
+
+/// Calculate the sample frequency and the total amount of samples.
+fn sample_config(period: Duration) -> (Duration, usize) {
+    let one_hour = Duration::from_secs(60.overflow_mul(60));
+    let base_interval = match period.cmp(&one_hour) {
+        Ordering::Less => period,
+        Ordering::Equal | Ordering::Greater => one_hour,
+    };
+    let divisor = 10;
+    let check_interval = base_interval
+        .checked_div(divisor)
+        .unwrap_or_else(|| unreachable!("duration divisor should not be 0"));
+    let check_interval_secs = check_interval.as_secs();
+    let periodic_secs = period.as_secs();
+    let length = periodic_secs
+        .overflow_div(check_interval_secs)
+        .overflow_add(1);
+    let retention =
+        usize::try_from(length).unwrap_or_else(|e| panic!("auto compact period is too large: {e}"));
+    (check_interval, retention)
+}
+
+#[async_trait::async_trait]
+impl<C: Compactable> Compactor for PeriodicCompactor<C> {
+    fn pause(&self) {
+        self.is_leader.store(false, Relaxed);
+    }
+
+    fn resume(&self) {
+        self.is_leader.store(true, Relaxed);
+    }
+
+    #[allow(clippy::integer_arithmetic)]
+    async fn run(&self) {
+        let mut last_revision: Option<i64> = None;
+        let shutdown_trigger = self.shutdown_trigger.listen();
+        let (sample_frequency, sample_total) = sample_config(self.period);
+        let mut ticker = tokio::time::interval(sample_frequency);
+        let mut revision_window = RevisionWindow::new(sample_total);
+        tokio::pin!(shutdown_trigger);
+        loop {
+            revision_window.sample(self.revision_getter.get());
+            tokio::select! {
+                _ = ticker.tick() => {
+                    if let Some(last_compacted_rev) = self.do_compact(last_revision, &revision_window).await {
```

Comment:
If a leader retires, a new leader is elected and do some compaction. After that the original leader is re-elected, will it try to compact the outdated revisions store in the buffer?

Reply:
No, the outdated revision will be overridden since the revision window keeps sampling every tick.

Reply:
What is the consequence of compacting the outdated revisions? The servers may have different durations for ticks.